### ISSUE TITLE
New derivation `builtin`

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nix#9774]
`builtins.derivation` (and `builtins.derivationStrict`) have a number of issues:

- `drvPath` has a funky "DrvDeep" string context that needs to be stripped away with the scary-sounding `builtins.unsafeDiscardOutputDependency`.

(TODO other things)

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by roberth** [comment]
We should align this with
- https://github.com/NixOS/nixpkgs/pull/281161

Currently that means that the return type should have a `path` field containing the `.drv`, but other than that, I've proposed to keep it lean.

**Comment by roberth** [comment]
We should also consider whether we even want to return such attributes as `inputSrcs`, `inputDrvs` eagerly, as this will probably result in plenty of unused-but-referenced `Value`s that we can't GC.
Technically we could work around that by adding them as thunks that call an (internal or not) primop for retrieving those from the `.drv` file. This is only a partial solution to the space leak though, because the attrs and thunks still take a bit of space, even if not evaluated.

**Comment by Ericson2314** [comment]
@roberth I was thinking the user would just drop them if they don't want them. But yes, this conflicts with `derivation` just being the result of the new primop.

**Comment by roberth** [comment]
Nix prints the position of the `name` attribute, but this is almost always useless:
```
whose name attribute is located at nixpkgs/pkgs/stdenv/generic/make-derivation.nix:331:7
```

Ideally, Nix would accept a `__pos` parameter with a user-provided position, such as `mkDerivation`'s `pos` argument, which is already used for some Nixpkgs-specific error messages. Nix should use this too, but `__pos` must not be written to the actual derivation in the store, as that would affect the hash every time the position changes slightly.
Changes to derivation attribute hashing need to be forward compatible, and we have no good forward compatibility in the current `derivation`/`derivationStrict` primops.

In summary:
- Implement measures for forward compatibility
- Use `__pos` argument for error positions, but don't write it to the derivation.

**Comment by roberth** [comment]
Added two examples of `meta` attributes that the store layer must be aware of.
